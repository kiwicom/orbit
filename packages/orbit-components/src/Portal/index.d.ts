// @flow
// Type definitions for @kiwicom/orbit-components
// Project: http://github.com/kiwicom/orbit-components

import * as React from "react";

declare module "@kiwicom/orbit-components/lib/Portal";

export interface Props {
  readonly renderInto?: string;
  readonly children: React.ReactNode;
}

declare const Portal: React.FunctionComponent<Props>;
export { Portal, Portal as default };
