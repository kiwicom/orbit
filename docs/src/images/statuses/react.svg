<svg width="25" height="24" viewBox="0 0 25 24" fill="none" xmlns="http://www.w3.org/2000/svg">
<path d="M22.75 11.9075C22.75 10.5826 21.0908 9.32695 18.547 8.54831C19.134 5.95556 18.8731 3.89278 17.7235 3.23237C17.4585 3.07746 17.1487 3.00408 16.8103 3.00408V3.91317C16.9979 3.91317 17.1487 3.94986 17.2751 4.01916C17.8295 4.33714 18.07 5.5479 17.8825 7.10518C17.8376 7.48838 17.7643 7.89197 17.6746 8.30371C16.8756 8.10803 16.0032 7.9572 15.0859 7.85936C14.5356 7.10518 13.9648 6.4203 13.39 5.82104C14.719 4.58581 15.9665 3.90909 16.8144 3.90909V3C15.6933 3 14.2257 3.79902 12.7418 5.18508C11.2579 3.80718 9.79036 3.01631 8.66928 3.01631V3.9254C9.51315 3.9254 10.7647 4.59804 12.0937 5.82511C11.5229 6.42438 10.9522 7.10518 10.41 7.85936C9.48869 7.9572 8.61629 8.10803 7.81727 8.30779C7.7235 7.90012 7.6542 7.50469 7.60528 7.12556C7.41368 5.56828 7.65012 4.35752 8.20047 4.03547C8.32277 3.96209 8.48176 3.92947 8.66928 3.92947V3.02038C8.32684 3.02038 8.01702 3.09376 7.74796 3.24868C6.60243 3.90909 6.3456 5.96779 6.93671 8.55238C4.40104 9.3351 2.75 10.5866 2.75 11.9075C2.75 13.2324 4.40919 14.488 6.95302 15.2666C6.36598 17.8594 6.62689 19.9221 7.7765 20.5826C8.04148 20.7375 8.35131 20.8108 8.69374 20.8108C9.81482 20.8108 11.2824 20.0118 12.7663 18.6258C14.2502 20.0037 15.7178 20.7945 16.8389 20.7945C17.1813 20.7945 17.4911 20.7212 17.7602 20.5662C18.9057 19.9058 19.1626 17.8471 18.5714 15.2625C21.099 14.4839 22.75 13.2283 22.75 11.9075ZM17.4422 9.18834C17.2914 9.71423 17.1039 10.2564 16.8919 10.7986C16.7247 10.4725 16.5494 10.1464 16.3578 9.82022C16.1703 9.49409 15.9705 9.17611 15.7708 8.86629C16.3497 8.9519 16.9082 9.05789 17.4422 9.18834ZM15.5751 13.53C15.2571 14.0803 14.931 14.6021 14.5926 15.0872C13.9852 15.1402 13.3696 15.1688 12.75 15.1688C12.1344 15.1688 11.5189 15.1402 10.9155 15.0913C10.5772 14.6062 10.2469 14.0885 9.92896 13.5422C9.61914 13.0082 9.33785 12.466 9.08102 11.9197C9.33378 11.3734 9.61914 10.8272 9.92489 10.2931C10.2429 9.74276 10.569 9.22095 10.9074 8.73583C11.5148 8.68284 12.1304 8.6543 12.75 8.6543C13.3656 8.6543 13.9811 8.68284 14.5845 8.73176C14.9228 9.21688 15.2531 9.73461 15.571 10.2809C15.8809 10.8149 16.1621 11.3571 16.419 11.9034C16.1621 12.4497 15.8809 12.9959 15.5751 13.53ZM16.8919 13C17.112 13.5463 17.2995 14.0925 17.4544 14.6225C16.9204 14.753 16.3578 14.863 15.7749 14.9486C15.9746 14.6347 16.1744 14.3127 16.3619 13.9825C16.5494 13.6563 16.7247 13.3261 16.8919 13ZM12.7582 17.3498C12.379 16.9584 11.9999 16.5222 11.6248 16.0453C11.9917 16.0616 12.3668 16.0738 12.7459 16.0738C13.1291 16.0738 13.5083 16.0656 13.8792 16.0453C13.5123 16.5222 13.1332 16.9584 12.7582 17.3498ZM9.72513 14.9486C9.14625 14.863 8.58775 14.757 8.05371 14.6266C8.20455 14.1007 8.39207 13.5585 8.60406 13.0163C8.7712 13.3424 8.94649 13.6686 9.1381 13.9947C9.3297 14.3208 9.52538 14.6388 9.72513 14.9486ZM12.7378 6.46514C13.1169 6.8565 13.496 7.2927 13.8711 7.76967C13.5042 7.75336 13.1291 7.74113 12.75 7.74113C12.3668 7.74113 11.9877 7.74929 11.6167 7.76967C11.9836 7.2927 12.3627 6.8565 12.7378 6.46514ZM9.72106 8.86629C9.5213 9.18019 9.32155 9.50224 9.13402 9.83245C8.94649 10.1586 8.7712 10.4847 8.60406 10.8108C8.38392 10.2646 8.19639 9.7183 8.04148 9.18834C8.57552 9.06197 9.1381 8.9519 9.72106 8.86629ZM6.0317 13.9702C4.58857 13.3547 3.65501 12.5475 3.65501 11.9075C3.65501 11.2674 4.58857 10.4562 6.0317 9.84468C6.38229 9.69384 6.76549 9.55932 7.16093 9.43294C7.39329 10.232 7.69904 11.0636 8.07817 11.9156C7.70312 12.7636 7.40145 13.5911 7.17316 14.3861C6.76957 14.2597 6.38636 14.1211 6.0317 13.9702ZM8.22493 19.7958C7.67051 19.4778 7.42998 18.267 7.61751 16.7097C7.66235 16.3265 7.73573 15.923 7.82542 15.5112C8.62444 15.7069 9.49684 15.8577 10.4141 15.9556C10.9644 16.7097 11.5352 17.3946 12.11 17.9939C10.781 19.2291 9.53353 19.9058 8.68559 19.9058C8.50214 19.9018 8.34723 19.8651 8.22493 19.7958ZM17.8947 16.6894C18.0863 18.2466 17.8499 19.4574 17.2995 19.7795C17.1772 19.8528 17.0182 19.8854 16.8307 19.8854C15.9869 19.8854 14.7353 19.2128 13.4063 17.9857C13.9771 17.3865 14.5478 16.7057 15.09 15.9515C16.0113 15.8537 16.8837 15.7028 17.6827 15.5031C17.7765 15.9148 17.8499 16.3102 17.8947 16.6894ZM19.4642 13.9702C19.1136 14.1211 18.7304 14.2556 18.335 14.382C18.1026 13.583 17.7969 12.7513 17.4178 11.8993C17.7928 11.0514 18.0945 10.2238 18.3228 9.42886C18.7264 9.55524 19.1096 9.69384 19.4683 9.84468C20.9114 10.4603 21.845 11.2674 21.845 11.9075C21.8409 12.5475 20.9074 13.3587 19.4642 13.9702Z" fill="#697D95"/>
<path d="M12.7458 13.771C13.7748 13.771 14.6089 12.9369 14.6089 11.908C14.6089 10.879 13.7748 10.0449 12.7458 10.0449C11.7169 10.0449 10.8828 10.879 10.8828 11.908C10.8828 12.9369 11.7169 13.771 12.7458 13.771Z" fill="#697D95"/>
</svg>
