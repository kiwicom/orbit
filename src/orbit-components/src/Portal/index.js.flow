// @flow
/*
  DOCUMENTATION: https://orbit.kiwi/components/portal/
*/

export type Props = {|
  +element?: string,
  +children: React$Node,
|};

declare export default React$ComponentType<Props>;
