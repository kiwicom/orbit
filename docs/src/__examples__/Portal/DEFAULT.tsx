import * as React from "react";
import {
  Button,
  Portal,
  Text,
  TextLink,
  Modal,
  ModalHeader,
  ModalSection,
} from "@kiwicom/orbit-components";

export default {
  Example: () => {
    const [showModal, setShowModal] = React.useState(false);
    return (
      <>
        <Portal>
          {showModal && (
            <Modal
              onClose={() => {
                setShowModal(false);
              }}
            >
              <ModalHeader title="More about portals" />
              <ModalSection>
                <Text>
                  For more on portals and how they work, read the{" "}
                  <TextLink href="https://reactjs.org/docs/portals.html" external>
                    React docs on portals.
                  </TextLink>
                </Text>
              </ModalSection>
            </Modal>
          )}
        </Portal>
        {!showModal && (
          <Button
            onClick={() => {
              setShowModal(true);
            }}
          >
            Show modal
          </Button>
        )}
      </>
    );
  },
  info: {
    title: "Default portal",
    description:
      "By default, portal inserts its children into a separate div created for the purpose. Events on a child still bubble to the React parent (even thought it's not a child in the DOM).",
  },
};
