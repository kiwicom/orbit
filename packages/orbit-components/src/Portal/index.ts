import * as React from "react";
import ReactDOM from "react-dom";

import { Props } from "./types";

const Portal = ({ renderInto, children }: Props) => {
  const [el] = React.useState(() => {
    if (typeof window !== "undefined") {
      return document.createElement("div");
    }
    return null;
  });

  const [node] = React.useState(() => {
    if (typeof window !== "undefined") {
      return renderInto && document.getElementById(renderInto)
        ? document.getElementById(renderInto)
        : document.body;
    }
    return null;
  });

  React.useLayoutEffect(() => {
    if (node && el) {
      node.appendChild(el);
    }
    return () => {
      if (node && el) {
        node.removeChild(el);
      }
    };
  }, [el, node]);

  if (typeof window !== "undefined" && el !== null) {
    return ReactDOM.createPortal(children, el);
  }

  return null;
};

export default Portal;
