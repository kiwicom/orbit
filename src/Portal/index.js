// @flow
import * as React from "react";
import ReactDOM from "react-dom";

import type { Props } from "./index";

export default class Portal extends React.Component<Props> {
  constructor(props: Props) {
    super(props);
    if (typeof window !== "undefined") {
      this.node =
        this.props.renderInto && document.getElementById(this.props.renderInto)
          ? document.getElementById(this.props.renderInto)
          : document.body;
      this.el = document.createElement("div");
    }
  }

  componentDidMount() {
    if (this.node && this.el) {
      this.node.appendChild(this.el);
    }
  }

  componentWillUnmount() {
    if (this.node && this.el) {
      this.node.removeChild(this.el);
    }
  }

  node: ?HTMLElement;

  el: ?HTMLElement;

  render() {
    const { children } = this.props;
    if (typeof window !== "undefined" && this.el) {
      return ReactDOM.createPortal(children, this.el);
    }
    return null;
  }
}
