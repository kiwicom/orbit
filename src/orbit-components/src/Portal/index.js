// @flow
import * as React from "react";
import ReactDOM from "react-dom";

import type { Props } from "./index";

export default class Portal extends React.Component<Props> {
  constructor(props: Props) {
    super(props);
    if (typeof window !== "undefined") {
      this.node =
        this.props.element && document.getElementById(this.props.element)
          ? document.getElementById(this.props.element)
          : document.body;
      this.el = document.createElement("div");
    }
  }

  componentDidMount() {
    if (this.node && this.el) {
      this.node.appendChild(this.el);
    }
  }

  componentWillUnmount() {
    if (this.node && this.el) {
      this.node.removeChild(this.el);
    }
  }

  node: ?HTMLElement;

  el: ?HTMLElement;

  render() {
    const { children } = this.props;
    if (typeof window !== "undefined" && this.el) {
      return ReactDOM.createPortal(children, this.el);
    }
    return null;
  }
}
