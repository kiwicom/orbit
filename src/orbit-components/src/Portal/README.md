# Portal
To implement Portal component into your project you'll need to add the import:
```jsx
import Portal from "@kiwicom/orbit-components/lib/Portal";
```
After adding import into your project you can use it simply like:
```jsx
<Portal>
  // probably Modal
  <Modal />
</Portal>
```
## Props
Table below contains all types of the props available in the Portal component.

| Name          | Type                  | Default         | Description                      |
| :------------ | :---------------------| :-------------- | :------------------------------- |
| children      | `React.Node`          |                 | The content of the Portal to render. [See Functional specs](#functional-specs)
| element       | `string`              | `"modal"`       | The string name of element to be used for the render.

## Functional specs
* `<Portal />` component will render passed children into a DOM node that exists outside the DOM hierarchy of the parent component. It means that you have to create `div` element with id `modal` (or whatever you want) in root `html` file. For more info, see [React Portal docs](https://reactjs.org/docs/portals.html).
