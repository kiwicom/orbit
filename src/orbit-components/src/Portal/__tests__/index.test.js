// @flow strict
import * as React from "react";
import { mount } from "enzyme";

import Portal from "../index";

describe("Portal with element", () => {
  const target = document.createElement("div");
  target.setAttribute("id", "portal");

  // $FlowIssue
  document.body.appendChild(target);
  const component = mount(
    <Portal element="portal">
      <p>Content</p>
    </Portal>,
  );
  it("should match snapshot", () => {
    expect(document.getElementById("portal")).toMatchSnapshot();
    expect(component).toMatchSnapshot();
  });
  it("unmouted, should match snapshot", () => {
    component.unmount();
    expect(document.getElementById("portal")).toMatchSnapshot();
    expect(component).toMatchSnapshot();
  });
});
describe("Portal without element", () => {
  const component = mount(
    <Portal>
      <p>Content</p>
    </Portal>,
  );
  it("should match snapshot", () => {
    expect(document.getElementById("modal")).toMatchSnapshot();
    expect(component).toMatchSnapshot();
  });
  it("unmouted, should match snapshot", () => {
    component.unmount();
    expect(document.getElementById("modal")).toMatchSnapshot();
    expect(component).toMatchSnapshot();
  });
});
