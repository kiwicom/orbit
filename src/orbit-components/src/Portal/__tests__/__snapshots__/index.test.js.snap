// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Portal with element should match snapshot 1`] = `
<div
  id="portal"
>
  <div>
    <p>
      Content
    </p>
  </div>
</div>
`;

exports[`Portal with element should match snapshot 2`] = `
<Portal
  element="portal"
>
  <Portal
    containerInfo={
      <div>
        <p>
          Content
        </p>
      </div>
    }
  >
    <p>
      Content
    </p>
  </Portal>
</Portal>
`;

exports[`Portal with element unmouted, should match snapshot 1`] = `
<div
  id="portal"
/>
`;

exports[`Portal with element unmouted, should match snapshot 2`] = `null`;

exports[`Portal without element should match snapshot 1`] = `null`;

exports[`Portal without element should match snapshot 2`] = `
<Portal>
  <Portal
    containerInfo={
      <div>
        <p>
          Content
        </p>
      </div>
    }
  >
    <p>
      Content
    </p>
  </Portal>
</Portal>
`;

exports[`Portal without element unmouted, should match snapshot 1`] = `null`;

exports[`Portal without element unmouted, should match snapshot 2`] = `null`;
